### PR TITLE
Added swd_uninit_debug() function. Used in nRF5x reset routine and flash uninit

### PR DESCRIPTION
**Pull request by gamnes**

### CODE REVIEW

**Comment by c1728p9** [comment]
@mbed-bot: TEST

File: source/daplink/interface/swd_host.c [hunk 1]
```diff
@@ -830,11 +830,31 @@ uint8_t swd_init_debug(void)
...
+    
+    return 1;
+}
+
 __attribute__((weak)) void swd_set_target_reset(uint8_t asserted)
 {
     (asserted) ? PIN_nRESET_OUT(0) : PIN_nRESET_OUT(1);
 }
 
+// This function is currently never called, see swd_set_target_state_sw() instead
```
Comment (gamnes) [inline source/daplink/interface/swd_host.c]: Ok, will remove comment

Comment (gamnes) [inline source/daplink/interface/swd_host.c]: Ok, will remove comment

File: source/target/nordic/target_reset_nrf5x_sam3u2c.c [hunk 2]
```diff
@@ -70,8 +70,11 @@ void swd_set_target_reset(uint8_t asserted)
                 swd_write_memory(0xE000ED0C, (uint8_t *)&swd_mem_write_data, 4);
                 //os_dly_wait(1);
             }
+            swd_uninit_debug();
         }
-        else {           
+        else {
+            swd_init_debug();
```
Comment (gamnes) [inline source/target/nordic/target_reset_nrf5x_sam3u2c.c]: No, but since we do uninit_debug() in the asserted statement above, we have to do init debug again here? Or can we always assume swd_set_target_reset() will be called with asserted == False after it has been called with asserted == True? Then i wouldnt have to call uninit_debug() in the asserted statement above, and then i wouldnt have to call init_debug in the NOTasserted statement here.

**Comment by gamnes** [comment]
Do you have the log of the failing mbed-bot testing?

**Comment by c1728p9** [comment]
It doesn't look like this was a real failure. I'll re-trigger.

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST

File: source/daplink/interface/swd_host.c [hunk 3]
```diff
@@ -878,6 +897,11 @@ uint8_t swd_set_target_state_hw(TARGET_RESET_STATE state)
             os_dly_wait(2);
             swd_set_target_reset(0);
             os_dly_wait(2);
+            
+            // Enable debug again because TARGET RESET might have disabled it
+            if (!swd_write_word(DBG_HCSR, DBGKEY | C_DEBUGEN)) {
+                return 0;
+            }
```
Comment (c1728p9) [inline source/daplink/interface/swd_host.c]: This line is causing tests to fail. The sequence here is to turn on the reset vector catch and then perform a hardware reset so the device is halted right at boot. By writing to the DHCSR here the C_HALT bit gets written to 0 causing the processor to start running. Is this change required for the NRF51 or NRF52 to work correctly?

File: source/target/nordic/target_reset_nrf5x_sam3u2c.c (hunk 2 above)
Comment (c1728p9) [inline source/target/nordic/target_reset_nrf5x_sam3u2c.c]: I'm not sure if that sequence can be guaranteed. What you have here is good.

File: source/daplink/interface/swd_host.c (hunk 3 above)
Comment (gamnes) [inline source/daplink/interface/swd_host.c]: Well, since its never used by the nRF51 nor the nRF52, this line is not necessary. It was added because the set_target_reset(0) might have disabled the debug connection. I can remove it.

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST
